File: java/src/Song.java
public class Song {
   private String name;
   private Song nextSong;

   public Song(String name) {
      this.name = name;
   }

   public void setNextSong(Song nextSong) {
      this.nextSong = nextSong;
   }

   public boolean isInRepeatingPlaylist() {
      //throw new UnsupportedOperationException("Waiting to be implemented.");

      // Empty and one element lists have no cycles:
      if (this == null || this.nextSong == null) {return false;}
      Song s1 = this; // slow
      Song s2 = this; // fast

      // whether there is a circle
      while (s2 != null && s2.nextSong != null) {
         s1 = s1.nextSong;
         s2 = s2.nextSong.nextSong;
         if (s1 == s2) {return true;}
      }
      return false;
   }

   public static void main(String[] args) {
      Song first = new Song("Hello");
      Song second = new Song("Eye of the tiger");

      first.setNextSong(second);
      second.setNextSong(first);

      System.out.println(first.isInRepeatingPlaylist());
   }
}

File: JavaScript/Song.js
class Song {
  name;
  nextSong;

  constructor(name) {
    this.name = name;
  }

  /**
   * @return {boolean} true if the playlist is repeating, false if not.
   */
  /*
  // Exceed Time Limit
  isInRepeatingPlaylist() {
    if (!this || !this.nextSong) {
      return false;
    }
    let slow = this;
    let fast = this;
    while (fast && fast.nextSong) {
      slow = slow.nextSong;
      fast = fast.nextSong.nextSong;
      if (slow === fast) {
        return true;
      }
    }

    return false;
  }
  */

  // use hashset
  isInRepeatingPlaylist() {
    let memo = new Set();
    let song = this;
    while (song) {
      if (memo.has(song)) {
        return true;
      } else {
        memo.add(song);
      }
      song = song.nextSong;
    }
    return false;
  }

  // use Object
  isInRepeatingPlaylist2() {
    let memo = new Object();
    let song = this;
    while (song) {
      if (memo[song.name]) {
        return true;
      } else {
        memo[song.name] = 1;
      }
      song = song.nextSong;
    }
    return false;
  }
}

let first = new Song('Hello');
let second = new Song('Eye of the tiger');

first.nextSong = second;
second.nextSong = first;

console.log(first.isInRepeatingPlaylist());
